### DISCUSSION TITLE
Rule editor components to support grouping of Rules

### DISCUSSION THREAD

**Discussion by justin0108** [gorules/zen#139] [General]
I have a lot of Rules (~100) to derive my final response.
Will there be a way to logically group a set of Rules and packed it under a component which can be tested independently?
Having this feature will allow:
- easier unit testing of smaller set of Rules
- easier grouping of related Rules
- final Rule flow will not be cluttered with too many boxes

**Comment by ivanmiletic** [comment]
Hi @justin0108 at the moment we do not have a plan for rule grouping. In our BRMS you can achieve this by using Decision Node (Reusable decision model).

Also, each model can be tested separately.

---
The "Decision" node is designed to extend the capabilities of decision models. Its function is to invoke and reuse other decision models during execution.

By incorporating the "Decision" node, developers can modularize decision logic, promoting reusability and maintainability in complex systems.

**Reply by justin0108** [reply]
Thanks for the prompt response. Manage to test out the concept using "Decision" component in BRMS

**Comment by justin0108** [comment]
Hi @ivanmiletic as mention in my original post I have around ~100 rules.
Is there any limit on the total number of DecisionNode that I can have in my final rule flow?
The ~100 rules is only the minimum and is set to grow up to a much larger number.

Assume all DecisionNode consist of DecisionTable, is there things which I should look out for?
- Rust engine memory limit?
- UI total nodes limit?
- any other limit?

I cannot find any performance benchmark on the GoRule website.

**Comment by ivanmiletic** [comment]
Hi @justin0108 ,

Performance wise there shouldn't be any issue in the single decision  model, we have tested with a few MB JDM file and it is extremely fast. Also, we have a concept In Memory loader that can load decision model in a memory (pre load) and just evaluate. It is available in all languages. Take a look at GO example here: https://github.com/gorules/zen-go/discussions/13#discussioncomment-8995476

Adding 100 nodes in UI should not be a problem performance wise, only issue is that it will probably become hard to manage that many nodes.

Would you be able to share what are you using? Are you using BRMS or Open Source, if so then which engine?

From BRMS 1.12.0 we have in mem caching available under the hood and it should be able to achieve high throughput such as 500+ req/s for simple decision models via http on a 1 vCPU machine (Azure/AWS) with 2gb of ram without sub decisions. BRMS is both horizontally and vertically scalable. For your use case the best would be to test.

**Comment by justin0108** [comment]
Hi @ivanmiletic 

Currently we are using IBM ODM with almost ~2000 rules.
Having 2000 nodes in a single UI is definitely not manageable. So your "Decision Node" in BRMS would come in really handy as it allows me to logically group related rules together. The final rule flow will be formed by chaining up all the "Decision Node".
I have not explore Memory loader yet but will definitely give it a try.

Is there significant performance penalty in having so many "Decision Node" (~100) since all these are external decision? Base on your docker image, I would assume in your BRMS the external decisions are loaded from Postgres during runtime.

**Reply by ivanmiletic** [reply]
Hi Justin, would you be able to send email to support@gorules.io or hi@gorules.io? It would be good to understand your use case a bit more, scale, performance, requirements (e.g. req/s).

As for your assumption that is correct, we do runtime loading and evaluation but we have built in in-mem caching for high loads (5s caching of the models by default). There might be small performance hit in cold start.

Outside of that we are working on additional enhancements to keep in memory models longer and do check with hashing in case of changes.
